docs/decisions/001X-python-client-constructors.md

```diff
@@ -0,0 +1,165 @@
+- Bad, because it is hard to keep track of all the clients and their parameters
+- Bad, because it is hard to discover the right client for a specific use case
+
+Example code:
+```python
+from agent_framework.openai import OpenAIClient # using a fictional OpenAIClient, to illustrate the point
+from agent_framework.azure import AzureOpenAIClient
+
+openai_client = OpenAIClient(api_key="...")
+azure_client = AzureOpenAIClient(deployment_name="...", ad_token_provider=...)
+```
+
+### Separate parameter set per backend with a single client, such as OpenAIClient with parameters, for endpoint/base_url, api_key, and entra auth.
+This option would entail a single client that can be used with different backends, but requires the user to pass in the right parameters.
+- Good, because it reduces the number of clients and makes it easier to discover the right client with the right parameters
+- Good, because it allows for a single client to be used with different backends and additional backends can be added easily
+- Good, because the user does not have to worry about which client to use, they can just use the `OpenAIClient` or `AnthropicClient` and pass in the right parameters, and we create the right client for them, if that client changes, then we do that in the code, without any changes to the api.
+- Good, because in many cases, the differences between the backends are just a few parameters, such as endpoint/base_url and authentication method.
+- Good, because client resolution logic could be encapsulated in a factory method, making it easier to maintain and even extend by users.
+- Bad, because it requires the user to know which parameters to pass in for the specific backend and when using environment variables, it is not always clear which parameters are used for which backend, or what the order of precedence is.
+- Bad, because it can lead to confusion if the user passes in the wrong parameters for the specific backend
+- Bad, because the name for a parameter that is similar but not the same between backends can be confusing, such as `deployment_name` for Azure OpenAI and `model_id` for OpenAI, would we then only have `model_id` for both, or have both parameters?
+- Bad, because it can lead to a lot of parameters that are not used for a specific backend, such as `entra_auth` for Azure OpenAI, but not for OpenAI
+- Bad, less stable per client, as changes to the parameter change all clients.
+- Bad, because customized behavior per backend is harder to implement, as it requires more conditional logic in the client code.
+
+Example code:
+```python
+from agent_framework.openai import OpenAIClient
+openai_client = OpenAIClient(
+    model_id="...",
+    api_key="...",
+)
+azure_client = OpenAIClient(
+    deployment_name="...", # or model_id, depending on the decision
+    ad_token_provider=...,
+)
+```
+
+### Single client with a explicit parameter for the backend to use, such as OpenAIClient(backend="azure") or AnthropicClient(backend="vertex").
```

Comment:
Looking at Anthropic as an example, they have separate SDK clients for Bedrock and Vertex. If a user wants to add support for a new backend that isn't officially supported, which option makes that easiest?